design/back.compatibility.md

```diff
@@ -0,0 +1,79 @@
+
+The obvious choice here is 2. Approach 1 would causes things to break often for no reason, no matter how careful developers are.
+
+### Source Code Management
+Here are 2 options to managing the source code
+1. move all the generation-specific code into a subfolder per generation - this is where the preview generation is worked on. When finalizing a generation and starting a new preview generation, create a copy of the preview generation folder. Change the namespace in the subfolder. Add a generation-specific prefix on each source file's filename
+2. move all generation-specific code into a separate repository, and have the main repository include this repository as a submodule for each generation.
+
+* Pros for 1: Easy to use with git. 
+* Cons for 1: It's clunky to have a prefix on filenames (but makes it much harder to accidentally modify things in a final generation - many tools highlight filename but not path). It's still possible to modify the wrong generation's files in the repo.
+* Pros for 2: Makes it harder to do accidental checkins in the wrong repo - you commit into a specific version.
+* Cons for 2: git submodule workflow is clunky, not integrated well with visual git tools. While it seems like meging across generations is easier, it's not something we'll see in practice as fixing finalized generations is not something we should be doing.
+
+I recommend version 1. The pain of the git submodule workflow, and loss of productivity is not worth the minor benefits of 2
+
+### Final Binaries
+Here are two options for building the the binaries
+1. Build a single DLL that includes all generations, with a small bit of generation-shared code for connecting to a server, finding the server version, and choosing the generation-specific mre. 
+2. Build one DLLs per generation, plus a management DLL. 
+Pros for 1: it's easier to have a single output DLL than multiple, as build scripts don't need to be updated when a generation is added. Total binary size is smaller than 2, as compiler will probably optimize out many functions that are unchanged across generations. Con: Higher build time for the DLL.
+I recommend option 1. 
+
+
+### Summary of Recommended Design
+We'll duplicate the files into a different subfolder per generation, add a generation-specific prefix to each file, and leave only the absolutely minimal set of files outside this folder. All of the code will be built into a single DLL.
+
+### Example
+For example, imagine the MRE SDK is at version 2.17.0. It would then support 3 generations of MREs: 1.0 (gen 1), 2.0 (gen 2), and 2.17.0 (pre-release for gen 3). 
+Any host app that includes this plugin will get permanent support for gen 1 and 2, and un-guaranteed prototype of gen3. If the host app later updates to SDK version 2.25.0, then gen 1 and gen 2 MREs run unchanged, but MREs built with gen3 version 2.17 to 2.24 may need to be updated 
```

Comment:
In this example, I think that MRE apps running with SDK 2.n.n (2.14, 2.24, etc) should map to the gen2 runtime. If we're developing a prerelease gen3 runtime, the corresponding SDK version should be major version 3, and include a prerelease tag: `3.0.0-prerelease.n`. This frees us up to patch the SDK with the intention of making it more stable against the existing runtime.

Reply:
Totally agree here.  It does not make sense to me that a 2.n.n is mapping to a release ahead.  Having the major release version number map to the generation version is far more understandable and natural in this case.  I don't know any software that does version numbering like this proposal, which already sets this up to be confusing and an outlier.  Usually the additional dot releases for a current version are bug fixes and stability improvements that do not introduce major features or breaking changes.  We should have the ability to patch a release for stability and increment the dot version without anyone running version 2 (gen 2) worrying about any breaking changes or issues.

Reply:
yeah as I wrote in another comment, I'm happy with that change. Don't know if it has to be 3.0.0-prerelease.n, 3.0.n, or 3.n.m-prerelease, but that's a detail based on what works for NPM, unity plugin packaging etc, I don't have strong feelings either way. I would like to make sure NPM and Unity versions are somewhat obviously linked.